src/emd/rt_bse.F

```diff
@@ -0,0 +1,1420 @@
+         CALL cp_cfm_to_cfm(rtbse_env%rho_new(i)%matrix, rtbse_env%rho(i)%matrix)
+      END DO
+
+      CALL timestop(handle)
+   END SUBROUTINE apply_delta_pulse
+! **************************************************************************************************
+!> \brief Determines the metric for the density matrix, used for convergence criterion
+!> \param rho_new Array of new density matrices (one for each spin index)
+!> \param rho_old Array of old density matrices (one for each spin index)
+!> \param nspin Number of spin indices
+!> \param workspace_opt Optionally provide external workspace to save some allocation time
+! **************************************************************************************************
+   FUNCTION rho_metric(rho_new, rho_old, nspin, workspace_opt) RESULT(metric)
+      TYPE(cp_cfm_p_type), DIMENSION(:), INTENT(IN)      :: rho_new, &
+                                                            rho_old
+      INTEGER, INTENT(IN)                                :: nspin
+      TYPE(cp_cfm_type), POINTER, OPTIONAL               :: workspace_opt
+      TYPE(cp_cfm_type)                                  :: workspace
+      REAL(kind=dp)                                      :: metric
+      REAL(kind=dp), DIMENSION(:), ALLOCATABLE           :: partial_metric
+      INTEGER                                            :: j
+      COMPLEX(kind=dp)                                   :: scale_factor
+
+      ALLOCATE (partial_metric(nspin))
+
+      ! Only allocate/deallocate storage if required
+      IF (PRESENT(workspace_opt)) THEN
+         workspace = workspace_opt
+      ELSE
+         CALL cp_cfm_create(workspace, rho_new(1)%matrix%matrix_struct)
+      END IF
+      scale_factor = 1.0
+      DO j = 1, nspin
+         CALL cp_cfm_to_cfm(rho_new(j)%matrix, workspace)
+         ! Get the difference in the resulting matrix
+         CALL cp_cfm_scale_and_add(scale_factor, workspace, -scale_factor, rho_old(j)%matrix)
+         ! Now, get the relevant number
+         partial_metric(j) = cp_cfm_norm(workspace, 'M')
+      END DO
+      metric = 0.0_dp
+      ! For more than one spin, do Cartesian sum of the different spin norms
+      DO j = 1, nspin
+         metric = metric + partial_metric(j)*partial_metric(j)
+      END DO
+      metric = SQRT(metric)
+      ! Deallocate workspace
+      IF (.NOT. PRESENT(workspace_opt)) CALL cp_cfm_release(workspace)
+      DEALLOCATE (partial_metric)
+   END FUNCTION
+
+! **************************************************************************************************
+!> \brief Determines the metric of the antihermitian part of the matrix
+!> \param real_fm Real part of the full matrix
+!> \param imag_fm Imaginary part of the full matrix
+! **************************************************************************************************
+   SUBROUTINE antiherm_metric(real_fm, imag_fm, workspace, metric)
+      TYPE(cp_fm_p_type), INTENT(IN)                    :: real_fm
```

Comment:
Compare comment on cp_cfm_p_type above

Reply:
Solved the same way as the comment above